File: readersWritersGo/resource.go
package main

import (
	"sync"

	log "github.com/Sirupsen/logrus"
)

// Resource definition
type Resource struct {
	content    string
	readSwitch Lightswitch
	roomEmpty  sync.Mutex
	turnstile  sync.Mutex
}

// NewResource create a Resource
func NewResource(initContent string) *Resource {
	return &Resource{content: initContent}
}

func (r *Resource) write(id int, data string) string {
	r.turnstile.Lock()
	r.roomEmpty.Lock()
	// critical section occurs here
	defer r.turnstile.Unlock()
	defer r.roomEmpty.Unlock()

	log.Info("--Writer entered the room ", id)
	defer log.Info("--Writer left the room ", id)

	//time.Sleep(100 * time.Millisecond)
	r.content = data
	log.Info("----Writer ", id, " wrote ", r.content)
	return r.content
}

func (r *Resource) read(id int) string {
	r.turnstile.Lock()
	r.turnstile.Unlock()
	r.readSwitch.Lock(&r.roomEmpty)
	// critical section occurs here
	defer r.readSwitch.Unlock(&r.roomEmpty)

	log.Info("--Reader entered the room ", id)
	defer log.Info("--Reader left the room ", id)

	//time.Sleep(100 * time.Millisecond)
	log.Info("----Reader ", id, " read ", r.content)
	return r.content
}

File: readersWritersJava/src/main/java/Resource.java
import java.util.concurrent.Semaphore;

public class Resource {

    private String content;
    private final Lightswitch readSwitch = new Lightswitch();
    private final Semaphore roomEmpty = new Semaphore(1);
    private final Semaphore turnstile = new Semaphore(1);

    public Resource(String initContent) {
        this.content = initContent;
    }

    public String write(int id, String data) throws InterruptedException {
        try {
            turnstile.acquire();
            roomEmpty.acquire();

            // critical section
            //System.out.printf("--Writer %d entered the room\n", id);

            //Thread.sleep(100);
            content = data;
            //System.out.printf("--Writer %d wrote '%s'\n", id, content);
            return content;
        } finally {
            turnstile.release();
            roomEmpty.release();
            //System.out.printf("--Writer %d left the room\n", id);
        }
    }

    public String read(int id) throws InterruptedException {
        try {
            turnstile.acquire();
            turnstile.release();
            readSwitch.lock(roomEmpty);

            // critical section
            //System.out.printf("--Reader %d entered the room\n", id);

            //Thread.sleep(100);
            //System.out.printf("--Reader %d read '%s'\n", id, content);
            return content;
        } finally {
            readSwitch.unlock(roomEmpty);
            //System.out.printf("--Reader %d left the room\n", id);
        }
    }
}